ENH: add option to return a sparse distance matrix in dijkstra

clane9 opened the pull request:
#### Reference issue

Closes gh-16585.

#### What does this implement/fix?

Adds an option to `csgraph.dijkstra` for returning a sparse distance matrix.

#### Additional information

I used libcpp vectors to construct the sparse distance matrix. As a result, `_shortest_path.pyx` now needs to be compiled with C++. I think I made the necessary changes to compile correctly. But it's possible I missed something.

I added new helper functions `_dijkstra_directed_sparse` and `_dijkstra_undirected_sparse` that are mostly copied from the originals. I also considered modifying the original helpers to compute a sparse representation by default. Lmk what approach you think is best.

Thanks in advance!

larsoner commented:
The idea seems reasonable to me!

The meson CI run being picky seems potentially legitimate:
```
scipy/sparse/csgraph/_shortest_path.cpython-39-x86_64-linux-gnu.so.p/_shortest_path.cpp:8124:108: error: ‘__pyx_v_dist_row’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
 8124 |     __pyx_t_3 = Py_BuildValue((char*) "("  __PYX_BUILD_PY_SSIZE_T  ")", ((Py_ssize_t)__pyx_v_dist_row->size())); if (unlikely(!__pyx_t_3)) __PYX_ERR(0, 650, __pyx_L1_error)
      |                                                                                      ~~~~~~~~~~~~~~~~~~~~~~^~
cc1plus: all warnings being treated as errors
```

clane9 commented:
Thanks for pointing that out, @larsoner! I'll see if I can sort it out.

lucascolley commented:
Hi @clane9, are you interested in returning to this? A start would be to merge main (if the conflicts are manageable) to see what CI thinks.

clane9 commented:
Hi @lucascolley, so sorry for abandoning this. I can try to pick it back up if it would still be useful. To start, I rebased with main to see what the current issues are.

lucascolley commented:
Great, thanks! I've approved the CI run so we can see what state this is in 👍

lucascolley commented:
CI has caught that you are comparing a `uint` with an `int`.

clane9 commented:
I removed the unused variables. Regarding the signed vs unsigned int comparisons, I'm a bit confused. It seems the complaints are about lines such as L374, where an `int` is compared to `uint` in the loop iteration.

https://github.com/scipy/scipy/blob/03dd1abece5988516b10d2bd75056ac3355c402e/scipy/sparse/csgraph/_shortest_path.pyx#L361-L379

But this code is outside my diff, so I'm not sure why the CI is complaining now? Would you want me to try to fix these errors?

rgommers commented on scipy/sparse/csgraph/meson.build:
```diff
@@ -25,6 +24,15 @@ foreach pyx_file: pyx_files
+  cython_gen_cpp.process('_shortest_path.pyx'),
+  cpp_args: cython_cpp_args,
+  include_directories: inc_np,
+  dependencies: [py3_dep],
```
The change here from `np_dep` to `py3_dep` does not look correct (`py3_dep` is always implicitly added already, and numpy is a dependency here). 

Also, `link_args` is dropped compared to the original code here.

rgommers commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -563,11 +575,27 @@ def dijkstra(csgraph, directed=True, indices=None,
         dist_matrix = np.full(N, np.inf, dtype=DTYPE)
         dist_matrix[indices] = 0
+    elif return_sparse_dist:
+        dist_matrix = np.empty((0, 0), dtype=DTYPE)
```
This should be `csr_matrix(np.empty(....))`, right?

rgommers reviewed (commented):
Thanks for this PR @clane9. 

It may be useful to add a small synthetic benchmark script, producing for example a plot showing the sub-N^2 scaling. I'm also curious when (matrix size, fill factor) this starts to help significantly.

> But this code is outside my diff, so I'm not sure why the CI is complaining now? Would you want me to try to fix these errors?

It looks like it's failing because you switched the file from C to C++. This causes Cython to emit warnings for some reason, and that fails the build in the CI job which adds `-Werror`.

More importantly, I don't quite understand the need yet for switching to C++. It looks like you are using `libcpp.vector` types where you could use regular memoryviews or 1-D ndarrays? I had a look at the original discussion, and it may have to do with dynamically expanding the array with `push_back`. However, since you're doing this element-by-element without bumping the vector capacity in larger steps, it's not clear to me how this would be more (or even equally) efficient to preallocating a large enough memoryview with in-place updates. Could you elaborate on this?

Regarding the API design and return types:
- It looks like the new keyword affects the `dist_matrix` return type but not the `predecessor_matrix` return type. Is that on purpose?
- The `min_only` interaction is a bit unfortunate. We now get two boolean keywords where one combination is disallowed to select between 3 return types. Looking ahead a bit, probably someone is going to want to change the return type from `csr_matrix` to `csr_array` at some point. That'd be return type number 4. It's worth thinking about whether there is a better alternative here.

clane9 commented:
Hi @rgommers, thanks for the helpful comments, much appreciated. I will work on addressing your comments but wanted to respond to a few things first.

> More importantly, I don't quite understand the need yet for switching to C++

I'm using the dynamically sized `libcpp.vector` to construct the sparse distance matrix because I don't know the number of non-zero entries in advance, and I want to avoid allocating an O(N^2) size array. I think you can see something similar being done in `ckdtree` [here](https://github.com/scipy/scipy/blob/466627abebdc91bf9376dfd823da378848a7696a/scipy/spatial/ckdtree/src/sparse_distances.cxx#L73). That said, I'm open to any better strategies. This was just the first I came up with.

> However, since you're doing this element-by-element without bumping the vector capacity in larger steps

My understanding was `libcpp.vector` expands exponentially, e.g. by 2x, whenever the capacity is reached. So I thought the reallocation cost would not be too severe. I agree it would be nice to preallocate a large enough array. But it's not clear to me how to bound the number of nonzeros.

> Regarding the API design and return types:

I agree with your points on the API design. I will think about this a bit more. But perhaps an option could be to change the flag from `return_sparse_dist` to `return_sparse`, change the behavior to return both a sparse distance and sparse predecessors, and clarify that in the case of `min_only=True` this flag has no effect (`dist_matrix` and `predecessors` are already vectors with shape `(n_nodes,)`, there's no need for them to be sparse).

As for what the sparse return type should be, whether `csr_matrix`, `csr_array`, or something else, I'm not picky and would vote for whatever makes most sense for the rest of the API. But I'd hope that we don't need more than one sparse return type.
